src/components/DepartmentsList.jsx

```diff
@@ -2,43 +2,158 @@ import React, { Component } from 'react';
+      window.scroll(0, 0)
+    }
   }
 
+  handlePagination = (direction) => {
+    this.setState((prev) => ({ currentPage: prev.currentPage + direction }));
+  };
+
+  addNewDepartment = (newDepartment) => {
+    this.setState((prev) => ({
+      departments: [...prev.departments, newDepartment],
+    }));
+  };
+
+  removeDepartment = (departmentId) => {
+    this.setState((prev) => ({
+      departments: prev.departments.filter(
+        (department) => department._id !== departmentId
+      ),
+    }));
+  };
+
+  updateDepartment = (updatedDepartment, departmentId) => {
+    const copiedState = this.state.departments;
+    const departmentIdx = copiedState.findIndex(
+      (department) => department._id === departmentId
+    );
+    if (departmentIdx !== -1) {
+      copiedState.splice(departmentIdx, 1, updatedDepartment);
+      this.setState({ departments: copiedState });
+    }
+  };
+
+  searchDepartment = (searchDepartments) => {
+    this.setState({ departments: searchDepartments });
+  };
+
+  sortDepartments = (sortedDepartments) => {
+    this.setState({ departments: sortedDepartments });
+  };
+
   render() {
-    const { departments } = this.state;
+    const { departments, lastPage, currentPage } = this.state;
+    const { handleOpenModal, handleCloseModal } = this.context;
     return (
       <section className="section">
         <div className="container section-wrap">
-          <AddNewItemBtn>department</AddNewItemBtn>
+          <div className="wrapper">
+            <button
+              onClick={() =>
+                handleOpenModal(
+                  <AddDepartmentForm
+                    add={this.addNewDepartment}
+                    close={handleCloseModal}
+                  ></AddDepartmentForm>
+                )
+              }
+              className="btn btn-success"
+            >
+              + Add department
+            </button>
+            <div className="form-wrap">
+              <SortForm sort={this.sortDepartments} />
+              <SearchForm search={this.searchDepartment} />
+            </div>
+          </div>
           <div className="item-list">
-          {departments &&
-            departments.map((department) => (
-              <Department key={department.id} department={department} />
-            ))}
+            {departments &&
+              departments.map((department) => (
+                <Department
+                  key={department._id}
+                  remove={this.removeDepartment}
+                  update={this.updateDepartment}
+                  department={department}
+                />
+              ))}
           </div>
+          {departments.length ? (
+            <>
+              {' '}
+              <span className='current-page'>Page: {currentPage}</span>
+              <div className="btn-wrap">
+                {currentPage > 1 && (
+                  <button
+                    onClick={() => this.handlePagination(-1)}
```

Comment:
handlePagination should accept the precise number of redirected page number, but not -1 -2 +10 etc